src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs

```diff
@@ -1580,6 +1667,55 @@ class C { }
             Assert.Equal(compilation, compilationsCalledFor[0]);
         }
 
+        [Fact]
+        public void IncrementalGenerator_Can_Add_Comparer_To_Join_Node()
+        {
+            var source = @"
+class C { }
+";
+            var parseOptions = TestOptions.Regular.WithLanguageVersion(LanguageVersion.Preview);
+            Compilation compilation = CreateCompilation(source, options: TestOptions.DebugDll, parseOptions: parseOptions);
+            compilation.VerifyDiagnostics();
+
+            Assert.Single(compilation.SyntaxTrees);
+
+            List<AdditionalText> texts = new List<AdditionalText>() { new InMemoryAdditionalText("abc", "") };
+
+            List<(Compilation, ImmutableArray<AdditionalText>)> calledFor = new List<(Compilation, ImmutableArray<AdditionalText>)>();
+
+            var generator = new IncrementalGeneratorWrapper(new PipelineCallbackGenerator(ctx =>
+            {
+                var compilationSource = ctx.Sources.Compilation.Join(ctx.Sources.AdditionalTexts)
+                                                // comparer that ignores the LHS (additional texts)
+                                                .WithComparer(new LambdaComparer<(Compilation, ImmutableArray<AdditionalText>)>((c1, c2) => c1.Item1 == c2.Item1, 0));
+                compilationSource.GenerateSource((spc, c) =>
+                {
+                    calledFor.Add(c);
+                });
+            }));
+
+            // run the generator once, and check it was passed the compilation + additional texts
+            GeneratorDriver driver = CSharpGeneratorDriver.Create(new ISourceGenerator[] { generator }, parseOptions: parseOptions, additionalTexts: texts);
+            driver = driver.RunGenerators(compilation);
+
+            Assert.Equal(1, calledFor.Count);
+            Assert.Equal(compilation, calledFor[0].Item1);
+            Assert.Equal(texts[0], calledFor[0].Item2.Single());
+
+            // edit the additional texts, and verify that the output was *not* called again on the next run
+            driver = driver.RemoveAdditionalTexts(texts.ToImmutableArray());
```

Comment:
>ToImmutableArray() [](http://example.com/codeflow?start=56&length=18)

It looks like `CSharpGeneratorDriver.Create()` takes an `IEnumerable<AdditionalText>` while `AddAdditionalTexts()` and `RemoveAdditionalTexts()` take an `ImmutableArray<AdditionalText>`. Is there a reason for the difference?

Reply:
`Create` is just a helper method, and calls `ToImmutableArray()` under the hood. We should probably expose a `Create` overload that takes an `ImmutableArray` directly, but this is the same as V1.